src/parser/shared/normalizers/Channeling.ts

```diff
@@ -0,0 +1,344 @@
+    // Warlock
+    buffChannelSpec(SPELLS.DRAIN_SOUL_TALENT.id),
+    // Priest
+    buffChannelSpec(SPELLS.DIVINE_HYMN_CAST.id),
+    nextCastChannelSpec(SPELLS.PENANCE_CAST.id),
+    buffChannelSpec(SPELLS.VOID_TORRENT_TALENT.id),
+    buffChannelSpec(SPELLS.MIND_FLAY.id), // TODO double check ID
+    buffChannelSpec(SPELLS.MIND_SEAR.id), // TODO double check ID
+    // Rogue
+    // Druid
+    buffChannelSpec(SPELLS.CONVOKE_SPIRITS.id),
+    // Monk
+    buffChannelSpec(SPELLS.ZEN_MEDITATION.id),
+    buffChannelSpec(SPELLS.ESSENCE_FONT.id),
+    buffChannelSpec(SPELLS.SOOTHING_MIST.id),
+    buffChannelSpec(SPELLS.CRACKLING_JADE_LIGHTNING.id),
+    buffChannelSpec(SPELLS.FISTS_OF_FURY_CAST.id),
+    // Demon Hunter
+    buffChannelSpec(SPELLS.EYE_BEAM.id), // TODO special handling because of the two buffs?
+    // Shaman
+    // Hunter
+    buffChannelSpec(SPELLS.RAPID_FIRE.id),
+    // TODO Barrage - can't find a single log with someone taking the talent...
+    // Paladin
+    // Warrior
+    buffChannelSpec(SPELLS.BLADESTORM_TALENT.id),
+    buffChannelSpec(SPELLS.BLADESTORM.id),
+    // Death Knight
+    buffChannelSpec(SPELLS.BLOODDRINKER_TALENT.id),
+  ];
+
+  // registered special case handlers, mapped by guid
+  channelSpecMap: { [key: number]: ChannelHandler } = {};
+
+  /**
+   * Constructs a Channeling normalizer that deals with normal hardcasts and instants by default,
+   * and then allows special case functions to be registered by spellId for channels.
+   * @param channelSpecs special case functions that handle specific spells
+   */
+  constructor(options: Options) {
+    super(options);
+
+    // populate the specs into a mapping for quick handling
+    Channeling.CHANNEL_SPECS.forEach((spec) => {
+      spec.guids.forEach((guid) => {
+        if (this.channelSpecMap[guid]) {
+          console.error(
+            'Channeling module trying to add more than one handler for spell guid ' +
+              guid +
+              ' - only the latest handler will apply!',
+          );
+        }
+        this.channelSpecMap[guid] = spec.handler;
+      });
+    });
+  }
+
+  normalize(events: AnyEvent[]) {
+    const eventsInserter: InsertableEventsWrapper = new InsertableEventsWrapper(events);
+
+    const channelState: ChannelState = {
+      unresolvedChannel: null,
+      eventsInserter,
+    };
+
+    // for each event, check if there is a special case handler for that GUID -
+    // if so - call the handler, if not - call the default handling
+    events.forEach((event: AnyEvent, index: number) => {
+      if (!HasSource(event) || event.sourceID !== this.selectedCombatant.id) {
+        return; // only concerned with events coming from selected player
+      }
+      const handler = HasAbility(event) ? this.channelSpecMap[event.ability.guid] : undefined;
+      if (handler) {
+        handler(event, events, index, channelState);
+      } else {
+        this.defaultHandler(event, events, index, channelState);
```

Comment:
is this going to trigger the default channel handler for non-channeling events?

Reply:
Yes. Judging by this comment and the one above, I think you have a misconception about what the `Channeling` module does (both in this rewrite and on live). Its purpose is to fabricate events that handle regular cast time spells and channeled spells in the same way, making it easier for users of these events to display when the player is casting and when he isn't.

It is the design intent of this module that regular cast time spells will have a `beginchannel` and `endchannel` event associated with them - this is laid out in the first paragraph of the class docs.

Cast time spells are pretty consistent in the original events out of WCL - a `begincast` at the start and a `cast` when it finishes - the purpose of the `defaultHandler` function is to handle these consistent baseline cases. Channeled spells are both different from cast time spells and also not internally consistent, hence the registered handlers for each of them.

Reply:
Ah, I'd missed that in the module description. While I think that having a consistent "actively casting/channeling" event marker is valuable, this currently breaks the timeline (non-channels are shown as channels, see below). 

![Screenshot from 2021-12-14 11-27-48](https://user-images.githubusercontent.com/4909458/146038952-31feb889-574d-4822-afa5-0ed7392ea2af.png)

Going to re-review the ABC changes and see if they require the change to channeling events. If so, I see a couple of options:

1. Add a prop to the channel events to indicate default vs non-default handler (not a fan of this).
2. Fabricate an "activitystart" / "activityend" event pair that is used for ABC instead of repurposing the "beginchannel" / "endchannel" pair.

If ABC doesn't require the change to casts having channeling events then I think we just don't emit channeling events for non-channels for the time being.

Reply:
Interestingly, I think the ABC code was broken for non-channel casts that were longer than the GCD prior to this. The GCD code doesn't appear to handle cast times (which would be a bad place to put it, but I needed to check) and the ABC code doesn't handle casts (only GCD and channel). This resulted in it being subtly wrong since few casts are longer than a GCD.

Another alternative is removing the CSS for channels. I don't know that they really need to be demarcated from casts, and the timeline suffers from high levels of visual noise that this could help with.

Reply:
![Screenshot from 2021-12-14 12-26-04](https://user-images.githubusercontent.com/4909458/146048789-f19e196c-c0d4-4bb7-8807-b1fa82bfd4f3.png)
actually prod looks like this right now. not sure why? fucking weird.

i think we just remove the channeling css and call it good. everything else seems fine.

Reply:
This is what I've been trying to say! I duplicated the behavior that's currently on prod. On prod, beginchannel and endchannel events are fabricated for regular hard casts, and those events are used by ABC to get active time right.  This is not a bug, this is intended behavior.

Reply:
okay i'm a dumdum. gonna merge this and we'll do the channeling timeline change later because wow that looks weird